feat(forms): add support for adding async validators via template

vsavkin opened the pull request:
Example:

```
@Directive({
  selector: '[uniq-login-validator]',
  providers: [provide(NG_ASYNC_VALIDATORS, {useExisting: UniqLoginValidator, multi: true})]
})
class UniqLoginValidator implements Validator {
  validate(c) { return someFunctionReturningPromiseOrObservable(); }
}
```

vicb commented on modules/angular2/src/core/forms/directives/ng_control_name.ts:
```diff
@@ -85,17 +91,16 @@ export class NgControlName extends NgControl implements OnChanges,
                   /* Array<Validator|Function> */ any[],
               @Optional() @Inject(NG_VALUE_ACCESSOR) valueAccessors: ControlValueAccessor[]) {
     super();
     this._parent = parent;
```
nit: could you update the ctor signature and drop this line ?

vsavkin replied:
Fixed

vicb commented on modules/angular2/src/core/forms/form_builder.ts:
```diff
@@ -34,7 +34,7 @@ import * as modelModule from './model';
  *       passwordRetry: builder.group({
  *         password: ["", Validators.required],
- *         passwordConfirmation: ["", Validators.required]
+ *         passwordConfirmation: ["", Validators.required, asyncValidation]
```
Validation -> Validator ?

vsavkin replied:
Fixed

vicb commented on modules/angular2/src/core/forms/validators.ts:
```diff
@@ -76,14 +78,33 @@ export class Validators {
+}
+
+function _executeValidators(control: modelModule.AbstractControl, validators: Function[]): any[] {
+  return validators.map(v => isPresent(v) ? v(control) : null);
```
do we need `isPresent()`  here ? (the list should be filtered already)

vsavkin replied:
Fixed

vicb commented on modules/angular2/test/core/forms/directives_spec.ts:
```diff
@@ -54,6 +56,19 @@ class CustomValidatorDirective implements Validator {
+    var completer = PromiseWrapper.completer();
+    var res = c.value != expected ? {"async": true} : null;
+    if (timeout == 0) {
+      PromiseWrapper.scheduleMicrotask(() => { completer.resolve(res); });
```
q: why do we need a microtask here ?

vicb commented:
LGTM

vsavkin committed:
feat(facade): add ObservableWrapper.fromPromise

vsavkin committed:
feat(forms): implements a combinator for composing async validators

vsavkin committed:
feat(forms): add support for adding async validators via template

Example:

@Directive({
  selector: '[uniq-login-validator]',
  providers: [provide(NG_ASYNC_VALIDATORS, {useExisting: UniqLoginValidator, multi: true})]
})
class UniqLoginValidator implements Validator {
  validate(c) { return someFunctionReturningPromiseOrObservable(); }
}

vsavkin committed:
feat(forms): update FormBuilder to support async validations

mary-poppins commented:
Merging PR #5020 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-5020](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-5020).
